provider/oci/provider.go

```diff
@@ -0,0 +1,336 @@
+}
+
+// ConfigDefaults implements config.ConfigSchemaSource
+func (o *EnvironProvider) ConfigDefaults() schema.Defaults {
+	return configDefaults
+}
+
+// Version implements environs.EnvironProvider.
+func (e EnvironProvider) Version() int {
+	return 0
+}
+
+// CloudSchema implements environs.EnvironProvider.
+func (e EnvironProvider) CloudSchema() *jsonschema.Schema {
+	return cloudSchema
+}
+
+// Ping implements environs.EnvironProvider.
+func (e *EnvironProvider) Ping(endpoint string) error {
+	return errors.NotImplementedf("Ping")
+}
+
+func validateCloudSpec(c environs.CloudSpec) error {
+	if err := c.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+	if c.Credential == nil {
+		return errors.NotValidf("missing credential")
+	}
+	if authType := c.Credential.AuthType(); authType != cloud.HTTPSigAuthType {
+		return errors.NotSupportedf("%q auth-type", authType)
+	}
+	return nil
+}
+
+// PrepareConfig implements environs.EnvironProvider.
+func (e EnvironProvider) PrepareConfig(args environs.PrepareConfigParams) (*config.Config, error) {
+	if err := validateCloudSpec(args.Cloud); err != nil {
+		return nil, errors.Annotate(err, "validating cloud spec")
+	}
+	// TODO(gsamfira): Set default block storage backend
+	return args.Config, nil
+}
+
+// Open implements environs.EnvironProvider.
+func (e *EnvironProvider) Open(params environs.OpenParams) (environs.Environ, error) {
+	logger.Infof("opening model %q", params.Config.Name())
+
+	if err := validateCloudSpec(params.Cloud); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	creds := params.Cloud.Credential.Attributes()
+	provider := providerCommon.NewJujuConfigProvider(
+		creds["user"], creds["tenancy"],
+		[]byte(creds["key"]), creds["fingerprint"],
+		creds["pass-phrase"], creds["region"])
+
+	client, err := providerCommon.NewOciClient(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	env := &Environ{
+		cli:   client,
+		p:     e,
+		clock: clock.WallClock,
+	}
+
+	if err := env.SetConfig(params.Config); err != nil {
+		return nil, err
+	}
+
+	env.namespace, err = instance.NewNamespace(env.Config().UUID())
+
+	cfg := env.ecfg()
+	if cfg.compartmentID() == nil {
+		return nil, errors.New("compartment-id may not be empty")
+	}
+
+	return env, nil
+}
+
+// CredentialSchemas implements environs.ProviderCredentials.
+func (e EnvironProvider) CredentialSchemas() map[cloud.AuthType]cloud.CredentialSchema {
+	return credentialSchema
+}
+
+func validateKey(key []byte, passphrase string) error {
+	keyBlock, _ := pem.Decode(key)
+	if keyBlock == nil {
+		return errors.Errorf("invalid private key")
+	}
+
+	if x509.IsEncryptedPEMBlock(keyBlock) {
+		if _, err := x509.DecryptPEMBlock(keyBlock, []byte(passphrase)); err != nil {
+			return errors.Annotatef(err, "decrypting private key")
+		}
+	}
+
+	return nil
+}
+
+// DetectCredentials implements environs.ProviderCredentials.
+func (e EnvironProvider) DetectCredentials() (*cloud.CloudCredential, error) {
```

Comment:
This is certainly a user interaction point that would want reasonably good testing and at least links out to what the definition of the OCI config files are.

Can you at least include URLs, possibly just to the OCI content that defines the structure of the INI file?

Reply:
yes. Adding link to documentation.